### PR TITLE
Parallelize `tox` invocations

### PR DESCRIPTION
**Pull request by portante**
Parallelize the python & legacy unit tests

We add the use of `pytest-xdist`, which allows us to pick the level of parallelism via the `-n` switch.  By default we use `-n auto` where it determines the parallelism to use based on the number of CPUs.

Use `parallel` to run all the legacy tests.  Each agent legacy sub-test is run as a separate parallel job, and then all the server tests are individually run through GNU `parallel`.

To facilitate the use of `parallel` for the server legacy tests, we break out the core server test harness out into a separate script. This helps to isolate the way tests are run from how the parallelism is achieved.

We also take the opportunity to use `local` in the various `bash` scripts, and make variable references consistent.

Finally, we address `mkdir -p` usage to remove its use, where possible, to ensure setup was performed without any hidden, or unexpected, directory creation, while fixing a problem in the server unit tests where the `common` directory was created but not used.

### CODE REVIEW

**Review by dbutenhof — COMMENTED** [review]
A lot to unpack here, but I'll drop some initial comments and look again later...

**Review by portante — COMMENTED** [review]
Updated with a rebase on PR #2551, which in turn, is based on PR #2663 now.  So Most of the files being changed here are from those PRs.  Once those two PRs land, and this PR is rebase directly on `main`, we'll be ready for a review.

The PR will still have multiple commits in order to facilitate making the reviews "easier".

**Comment by portante** [comment]
Now that PR #2551 and #2663 were merged, we have one other PR we need merged first, PR #2682.

**Review by portante — COMMENTED** [review]
One #2682 lands we'll be ready for this PR to be reviewed.

**Review by portante — COMMENTED** [review]
Ready for review!

File: exec-unittests
Comment (dbutenhof) [inline exec-unittests:136]: Yikes!

> --citation
> Print the citation notice and BibTeX entry for GNU parallel, silence citation notice for all future runs, and exit. It will not run any commands.
> 
> If it is impossible for you to run --citation you can instead use --will-cite, which will run commands, but which will only silence the citation notice for this single run.
> 
> If you use --will-cite in scripts to be run by others you are making it harder for others to see the citation notice. The development of GNU parallel is indirectly financed through citations, so if your users do not know they should cite then you are making it harder to finance development. However, if you pay 10000 EUR, you have done your part to finance future development and should feel free to use --will-cite in scripts.
> 
> If you do not want to help financing future development by letting other users see the citation notice or by paying, then please consider using another tool instead of GNU parallel. You can find some of the alternatives in man parallel_alternatives.
> 

Can we use `--will-cite`??

**Review by dbutenhof — APPROVED** [review]
The massive code relocations are difficult to review, but I'm assuming it's really just moved and I didn't review those commits line by line.

The GNU parallel "citation" adware is annoying, but that's kinda GNU; would the documented `--will-cite` be cleaner than manually touching a file you presumably expect it to use to track this?

File: exec-unittests
Comment (webbnh) [inline exec-unittests:59]: Is `_posargs` being used?  (Should it be?  🙂)

File: exec-unittests
Comment (webbnh) [inline exec-unittests:160]: I think we should be setting `rc` to `1`, here -- aside from the fact that we probably don't care _how many_ jobs failed, the fact that it might be overwritten with a different, subsequent failure value means that we cannot count on it being representative of more than a single bit of resolution.

File: server/bin/unittests
Comment (webbnh) [inline server/bin/unittests:341]: Perhaps
```
_para_dir=${HOME:-/tmp/}.parallel
```

File: server/bin/unittests
Comment (webbnh) [inline server/bin/unittests:365]: I think that, if `$?` is non-zero, we should set `failures` to `1` rather than copying the exit value of `parallel`.

File: server/bin/utils/run-unittest
Comment (webbnh) [inline server/bin/utils/run-unittest:585]: I think `$(<` is preferable to `$(cat`.

Ditto for lines 587 and 601.

File: server/bin/utils/run-unittest
Comment (webbnh) [inline server/bin/utils/run-unittest:591]: Do you really want all of these in separate commands?  Given that this is the `PASS` case, wouldn't `rm -rf ${_testroot}` do what we want?  If not, can't we combine the first three into a single command?...why does the first omit the `-f`?

File: server/bin/utils/run-unittest
Comment (webbnh) [inline server/bin/utils/run-unittest:603]: What about `_testres` and/or `_test_db`?

**Review by webbnh — DISMISSED** [review]
Just small stuff.

File: exec-unittests
Comment (webbnh) [reply exec-unittests:136]: I feel like we should expand this code comment to point out that this usage of `parallel` is internal and automated; that we are not distributing product code which depends on this tool; and that we, as developers, have viewed the citation and therefore are justified suppressing future displays of it in our development processes.

**Comment by portante** [comment]
FWIW, I kicked off a rebuild of the Jenkins job and it completed and updated the job status.

File: server/bin/unittests
Comment (portante) [reply server/bin/unittests:45]: This is existing code, but will add a comment.

File: server/bin/utils/run-unittest
Comment (portante) [reply server/bin/utils/run-unittest:21]: I was trying to avoid disturbing the gold files which fail when the extra space is added.

File: server/bin/utils/run-unittest
Comment (portante) [reply server/bin/utils/run-unittest:550]: Fixed in a later commit.

File: exec-unittests
Comment (portante) [reply exec-unittests:66]: We ran into this in the past, I believe.  If you have `shift 2` but don't have two arguments then the behavior is odd.

File: exec-unittests
Comment (portante) [reply exec-unittests:157]: You can't, which is what the comment refers to.  If you combine the loops the count is lost because it is executed in a sub-shell.

File: server/bin/utils/run-unittest
Comment (portante) [reply server/bin/utils/run-unittest:591]: We wanted to see when things were left around or something was missing.  In particular, the `rmdir` should be cleaning up an empty directory.

File: server/bin/utils/run-unittest
Comment (portante) [reply server/bin/utils/run-unittest:603]: Yes, so that one can go back and look at the data related to the failure.

File: exec-unittests
Comment (portante) [reply exec-unittests:115]: Removing this did not change test behavior and they all passed.

File: server/bin/utils/run-unittest
Comment (portante) [reply server/bin/utils/run-unittest:21]: Tests failed with white-space mismatch.  Would rather not fix that in this commit since this code is just a move of the existing behavior.

File: exec-unittests
Comment (dbutenhof) [reply exec-unittests:61]: > n must be a non-negative number less than or equal to $#.  If n is 0, no parameters are changed.  If n is not given, it is assumed  to  be  1.   _If  n  is greater than $#, the positional parameters are not changed._

File: exec-unittests
Comment (dbutenhof) [inline exec-unittests:138]: So did `--will-cite` not work? Is there documentation somewhere about creating this `will-cite` marker file? (I didn't find any, although `parallel` seems to try very hard to be opaque about suppressing the nagware.) Otherwise, it seems something like `alias parallel='parallel --will-cite'` would have been cleaner.

File: server/bin/unittests
Comment (dbutenhof) [reply server/bin/unittests:49]: Really? I'd looked through the old code but didn't find it. In any case, thanks. 👍🏻

**Review by dbutenhof — DISMISSED** [review]

File: server/bin/unittests
Comment (portante) [reply server/bin/unittests:49]: See https://github.com/distributed-system-analysis/pbench/blob/main/server/bin/unittests#L44.

File: exec-unittests
Comment (webbnh) [reply exec-unittests:61]: > If n is greater than $#, the positional parameters are not changed.

Oh, urk.

So, then...
```
    _major=${1}
    _subtst=${2}
    shift; shift
```
🤪

**Review by dbutenhof — APPROVED** [review]
Nice!

File: exec-unittests
Comment (webbnh) [reply exec-unittests]: Well, gosh!

OK, I propose more specific wording for the comment, then.  E.g., something like
```
        # Count what we _will_ do since the `for` command which drives the
        # `parallel` command below will be run in a sub-shell (because it is
        # part of a "pipeline") which will make changes to `count` unavailable.
```

Alternatively, we could build the list and count it in a single loop, and then invoke `parallel` separately:
```
        let count=0
        test_list=""
        for _subtst in ${_subtst_list}; do
            test_list+="${_major} ${_subtst} ${posargs}"$'\n'
            (( count++ ))
        done
        parallel --lb -k ${para_jobs_arg} para_run_legacy <<< ${test_list}
```

File: exec-unittests
Comment (webbnh) [inline exec-unittests:133]: s/disturbing/distributing/  😁

Also, I would change "to depend" to "which depends".

File: server/bin/utils/run-unittest
Comment (webbnh) [inline server/bin/utils/run-unittest:542]: As noted elsewhere, the `shift 2` is potentially problematic...should we replace it with `shift; shift`?

File: server/bin/utils/run-unittest
Comment (webbnh) [inline server/bin/utils/run-unittest:546]: How are we feeling today about the `-p`?  (I have no problem with it, but lately we've been removing it elsewhere.)

**Review by webbnh — APPROVED** [review]
Looks good; there are just nits remaining:
- in a couple of places, `shift 2` vs. `shift; shift` or separate shifts;
- some typos and other editorials in code comments; and,
- some other questions/suggestions.

File: server/bin/utils/run-unittest
Comment (portante) [reply server/bin/utils/run-unittest:542]: This is case it is not because `server/bin/unittests` always/must invoke `run-unittest` with the test name and the command to run, the additional arguments are optional.

File: server/bin/utils/run-unittest
Comment (portante) [reply server/bin/utils/run-unittest:543]: Okay, so because of this prompt, I looked at the other uses of `mkdir -p` and found that we were unnecessarily creating a `common` directory.  I've added a separate commit that addresses that along with removing the remaining uses of `mkdir -p` that hide errors.

**Review by dbutenhof — DISMISSED** [review]

**Review by dbutenhof — DISMISSED** [review]

**Review by webbnh — DISMISSED** [review]
I think you may have removed a `-p` which we would be better to keep, and I think the wording in the `parallel` comment should have a tweak...but, we can negotiate on those.  Also, I'm still wondering what your thoughts are on a simplification that I offered in the last round.

File: exec-unittests
Comment (webbnh) [reply exec-unittests:138]: Sounds good; so what do you think of the simplification I offered above?

File: server/bin/utils/run-unittest
Comment (portante) [reply server/bin/utils/run-unittest:248]: I believe the `-p` will hide unexpected directory creations that should have been properly setup.

And yes one can argue that the relationship is setup elsewhere, but it is really this script, and so I'll create a proper variable to encapsulate that.

**Review by dbutenhof — DISMISSED** [review]

**Review by dbutenhof — DISMISSED** [review]

**Review by portante — COMMENTED** [review]
Just removed the commits used to verify Jenkins integration was working.

**Review by dbutenhof — DISMISSED** [review]
Maybe for the final time? 😆

File: jenkins/branch.name
Comment (dbutenhof) [inline jenkins/branch.name:1]: OK, I guess we're *not* done yet...

**Review by webbnh — APPROVED** [review]
Good to go...except for the "DO NOT MERGE" commit.

**Comment by webbnh** [comment]
Oh, and, can we have a better description for the PR than "TBD"?  😀

**Review by webbnh — APPROVED** [review]
👍

**Review by dbutenhof — APPROVED** [review]

**Pull request merged**
